STYLE.md

```diff
@@ -0,0 +1,432 @@
+Haskell Style Guide
+============================
+> This style guide is mostly a copy of [Johan Tibell's guide](https://github.com/tibbe/haskell-style-guide/blob/master/haskell-style.md)
+> with some restructurization, elaboration on some topics and some additions.
+> This style guide's aims are code beauty, readability and understandability.
+
+General guide lines
+-------------------
+### Line Length
+
+Maximum line length is *80 characters* or *100 characters* if necessary.
+
+Modern screens have high definition and big width.
+But with some tiling managers and two terminals on one screen you are not able to
+see many characters on one line.
+On the other hand, restricting line size to a very small number like 80 leads to
+some crazy indentation despite the fact that
+shorter lines should force you to write well structured code.
+That's why *100* is a reasonable compromize.
+
+
+### Indentation
+
+Tabs are illegal.  Use spaces for indenting.  Indent your code blocks
+with *2 spaces*.  Indent the `where` keyword with two spaces to set it
+apart from the rest of the code and indent the definitions in a
+`where` clause with 2 spaces.  Some examples:
+
+```haskell
+sayHello :: IO ()
+sayHello = do
+    name <- getLine
+    putStrLn $ greeting name
+  where
+    greeting name = "Hello, " ++ name ++ "!"
+
+filter :: (a -> Bool) -> [a] -> [a]
+filter _ [] = []
+filter p (x:xs)
+  | p x       = x : filter p xs
+  | otherwise = filter p xs
+```
+
+### Blank Lines
+
+* One blank line between top-level definitions.
+* No blank lines between type signatures and function definitions.
+* Add one blank line between
+functions in a type class instance declaration if the function bodies
+are large.
+* You can add blank lines inside a big `do` block to separate logical
+parts of it.
+* You can also use blank lines to separate definitions inside `where`
+clause, if necessary.
+
+### Whitespace
+
+* Surround binary operators with a single space on either side. In case of
+currying add one space between the argument and the operation.
+
+* Use some tool to remove trailing spaces.
+
+### Naming convention
+
+Casing:
++ **_lowerCamelCase_** for function and variable names.
++ **_UpperCamelCase_** for types.
+
+* Don't use short names like `n`, `sk`, `f` unless their meaning is clear from
+context (function name, types, other variables, etc.).
+
+* For readability reasons, don't capitalize all letters when using an
+abbreviation.  For example, write `HttpServer` instead of
+`HTTPServer`.  Exception: two or three letter abbreviations, e.g. `IO`, `STM`.
+
+**Records name conventions**
+
+If data type has only one constructor then this data type name should be same
+as constructor name (also applies to `newtype`).
```

Comment:
Counter examples in core code that I've seen only just recently:

- https://github.com/input-output-hk/cardano-ledger/blob/master/cardano-ledger/src/Cardano/Chain/UTxO/Tx.hs#L51
- https://github.com/input-output-hk/cardano-ledger/blob/master/cardano-ledger/src/Cardano/Chain/UTxO/Tx.hs#L123

Reply:
This is a good point, making changes.

Reply:
Updated.

Reply:
The `Tx = UnsafeTx` almost certainly came from the Serokell code. There is not reason to perpretuate Serokell's insanity in the new code base. If I constructor actually is unsafe maybe it could be private to the module or some other approach taken.

Reply:
@erikd  `UnsafeTx` exists in `cardano-ledger`: https://github.com/input-output-hk/cardano-ledger/blob/583718d76bd9952336c588f9c48828756ba83254/cardano-ledger/src/Cardano/Chain/UTxO/Tx.hs#L51

Hiding the constructor did occur to me but I wasn't sure which was better/best practice as `UnsafeTx` was in `cardano-ledger`.

Reply:
Hiding a constructor that can be misused is a good idea (if possible).

Reply:
@erikd agree that saying `data Tx = Tx`, not exporting the data constructor, and exporting a smart constructor would be a good approach.